File: MindLinkApp/MindLinkApp/Processing/SignalProcessor.swift
import Foundation
import Accelerate

// MARK: - Band Powers Result

struct BandPowers {
    var delta: Double = 0
    var theta: Double = 0
    var alpha: Double = 0
    var beta:  Double = 0
    var gamma: Double = 0

    var ratios: BandRatios {
        BandRatios(
            alphaBeta:       beta  > 0 ? alpha / beta  : 0,
            thetaAlpha:      alpha > 0 ? theta / alpha : 0,
            betaDelta:       delta > 0 ? beta  / delta : 0,
            alphaThetaBeta:  beta  > 0 ? (alpha + theta) / beta : 0
        )
    }
}

struct BandRatios {
    var alphaBeta:      Double = 0
    var thetaAlpha:     Double = 0
    var betaDelta:      Double = 0
    var alphaThetaBeta: Double = 0
}

// MARK: - SignalProcessor (port of signal_processor.py)

class SignalProcessor {

    private let sampleRate: Double
    private var rawBuffer: [Double] = []
    private let bufferSize = 512

    init(sampleRate: Double = Config.sampleRate) {
        self.sampleRate = sampleRate
    }

    // MARK: - Add raw EEG sample

    func addRawSample(_ value: Double) {
        rawBuffer.append(value)
        if rawBuffer.count > bufferSize { rawBuffer.removeFirst() }
    }

    // MARK: - Process raw buffer → frequency bands

    func processRawEEG(_ rawValue: Double? = nil) -> BandPowers {
        if let v = rawValue { addRawSample(v) }
        guard rawBuffer.count >= Config.windowSize else { return BandPowers() }

        var signal = Array(rawBuffer.suffix(Config.windowSize))

        // Remove DC
        let mean = signal.reduce(0, +) / Double(signal.count)
        signal = signal.map { $0 - mean }

        // Hann window to reduce spectral leakage
        var window = [Double](repeating: 0, count: signal.count)
        vDSP_hann_windowD(&window, vDSP_Length(signal.count), Int32(vDSP_HANN_NORM))
        vDSP_vmulD(signal, 1, window, 1, &signal, 1, vDSP_Length(signal.count))

        return extractBands(from: signal)
    }

    // MARK: - Process TGAM1 pre-computed band powers (from TGAM1 chip)

    func processTGAMPowerData(_ eegPower: [String: Double]) -> BandPowers {
        let total = eegPower.values.reduce(0, +)
        guard total > 0 else { return BandPowers() }

        return BandPowers(
            delta: (eegPower["delta"] ?? 0) / total,
            theta: (eegPower["theta"] ?? 0) / total,
            alpha: ((eegPower["low_alpha"] ?? 0) + (eegPower["high_alpha"] ?? 0)) / total,
            beta:  ((eegPower["low_beta"]  ?? 0) + (eegPower["high_beta"]  ?? 0)) / total,
            gamma: ((eegPower["low_gamma"] ?? 0) + (eegPower["high_gamma"] ?? 0)) / total
        )
    }

    // MARK: - FFT + Band Extraction

    private func extractBands(from samples: [Double]) -> BandPowers {
        let n = samples.count
        guard n >= 2 else { return BandPowers() }

        // Next power of 2 for FFT
        let fftSize = nextPowerOf2(n)
        var padded = samples + [Double](repeating: 0, count: fftSize - n)

        // Split-complex FFT via vDSP
        let log2n = vDSP_Length(log2(Double(fftSize)))
        guard let setup = vDSP_create_fftsetupD(log2n, FFTRadix(kFFTRadix2)) else {
            return BandPowers()
        }
        defer { vDSP_destroy_fftsetupD(setup) }

        var splitR = [Double](repeating: 0, count: fftSize / 2)
        var splitI = [Double](repeating: 0, count: fftSize / 2)
        padded.withUnsafeBytes { ptrRaw in
            let ptr = ptrRaw.bindMemory(to: DSPDoubleComplex.self)
            var split = DSPDoubleSplitComplex(realp: &splitR, imagp: &splitI)
            vDSP_ctozD(ptr.baseAddress!, 2, &split, 1, vDSP_Length(fftSize / 2))
            vDSP_fft_zripD(setup, &split, 1, log2n, FFTDirection(kFFTDirection_Forward))
        }

        // Compute power spectrum
        var power = [Double](repeating: 0, count: fftSize / 2)
        for i in 0 ..< fftSize / 2 {
            power[i] = splitR[i] * splitR[i] + splitI[i] * splitI[i]
        }
        let totalPower = power.reduce(0, +)
        guard totalPower > 0 else { return BandPowers() }

        let freqRes = sampleRate / Double(fftSize)   // Hz per bin

        func bandPower(low: Double, high: Double) -> Double {
            let loIdx = Int(low  / freqRes)
            let hiIdx = min(Int(high / freqRes), fftSize / 2 - 1)
            guard loIdx <= hiIdx else { return 0 }
            return power[loIdx...hiIdx].reduce(0, +) / totalPower
        }

        return BandPowers(
            delta: bandPower(low: 0.5,  high: 4.0),
            theta: bandPower(low: 4.0,  high: 8.0),
            alpha: bandPower(low: 8.0,  high: 13.0),
            beta:  bandPower(low: 13.0, high: 30.0),
            gamma: bandPower(low: 30.0, high: 100.0)
        )
    }

    func reset() { rawBuffer.removeAll() }

    // MARK: - Helpers

    private func nextPowerOf2(_ n: Int) -> Int {
        var p = 1
        while p < n { p <<= 1 }
        return p
    }
}

File: signal_processor.py
"""
Signal processing module for EEG frequency band extraction
Processes RAW EEG data to extract Alpha, Beta, Theta, Delta, and Gamma bands
"""

import numpy as np
from scipy import signal
from scipy.fft import fft, fftfreq
from typing import Dict, Optional, Tuple
import logging
from config import SAMPLE_RATE, WINDOW_SIZE, OVERLAP, FREQUENCY_BANDS

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class SignalProcessor:
    """
    Processes RAW EEG signals to extract frequency bands and features
    """
    
    def __init__(self, sample_rate: int = SAMPLE_RATE):
        self.sample_rate = sample_rate
        self.raw_buffer = []
        self.buffer_size = 512  # Keep last 512 samples
        self.frequency_bands = FREQUENCY_BANDS
        
    def add_raw_sample(self, raw_value: float):
        """Add a raw EEG sample to the buffer"""
        self.raw_buffer.append(raw_value)
        if len(self.raw_buffer) > self.buffer_size:
            self.raw_buffer.pop(0)
    
    def extract_frequency_bands(self, signal_data: np.ndarray) -> Dict[str, float]:
        """
        Extract power in different frequency bands using FFT
        Args:
            signal_data: Array of EEG samples
        Returns:
            Dictionary with power values for each frequency band
        """
        if len(signal_data) < WINDOW_SIZE:
            return self._get_empty_bands()
        
        # Apply window function to reduce spectral leakage
        windowed = signal_data * signal.windows.hann(len(signal_data))
        
        # Compute FFT
        fft_values = fft(windowed)
        fft_freqs = fftfreq(len(windowed), 1.0 / self.sample_rate)
        
        # Get power spectral density
        power_spectrum = np.abs(fft_values) ** 2
        
        # Extract power in each frequency band
        bands = {}
        total_power = np.sum(power_spectrum)
        
        for band_name, (low_freq, high_freq) in self.frequency_bands.items():
            # Find indices for frequency range
            band_mask = (fft_freqs >= low_freq) & (fft_freqs <= high_freq)
            band_power = np.sum(power_spectrum[band_mask])
            
            # Normalize by total power
            bands[band_name] = band_power / total_power if total_power > 0 else 0.0
        
        return bands
    
    def process_raw_eeg(self, raw_value: Optional[float] = None) -> Dict[str, float]:
        """
        Process raw EEG value and extract frequency bands
        Args:
            raw_value: Raw EEG sample value
        Returns:
            Dictionary with frequency band powers
        """
        if raw_value is not None:
            self.add_raw_sample(raw_value)
        
        if len(self.raw_buffer) < WINDOW_SIZE:
            return self._get_empty_bands()
        
        # Convert buffer to numpy array
        signal_array = np.array(self.raw_buffer[-WINDOW_SIZE:])
        
        # Remove DC component
        signal_array = signal_array - np.mean(signal_array)
        
        # Apply bandpass filter (0.5-100 Hz) to remove noise
        try:
            sos = signal.butter(4, [0.5, 100], btype='band', 
                              fs=self.sample_rate, output='sos')
            filtered_signal = signal.sosfilt(sos, signal_array)
        except:
            filtered_signal = signal_array
        
        # Extract frequency bands
        bands = self.extract_frequency_bands(filtered_signal)
        
        return bands
    
    def process_tgam_power_data(self, eeg_power: Dict) -> Dict[str, float]:
        """
        Process TGAM1's pre-computed EEG power values
        TGAM1 provides: delta, theta, low_alpha, high_alpha, 
                        low_beta, high_beta, low_gamma, high_gamma
        """
        # Combine bands according to our frequency definitions
        total_power = sum(eeg_power.values())
        
        if total_power == 0:
            return self._get_empty_bands()
        
        bands = {
            'delta': eeg_power.get('delta', 0) / total_power,
            'theta': eeg_power.get('theta', 0) / total_power,
            'alpha': (eeg_power.get('low_alpha', 0) + 
                     eeg_power.get('high_alpha', 0)) / total_power,
            'beta': (eeg_power.get('low_beta', 0) + 
                    eeg_power.get('high_beta', 0)) / total_power,
            'gamma': (eeg_power.get('low_gamma', 0) + 
                     eeg_power.get('high_gamma', 0)) / total_power
        }
        
        return bands
    
    def calculate_band_ratios(self, bands: Dict[str, float]) -> Dict[str, float]:
        """
        Calculate important frequency band ratios for fatigue detection
        """
        ratios = {}
        
        # Alpha/Beta ratio (indicates relaxation vs alertness)
        if bands.get('beta', 0) > 0:
            ratios['alpha_beta'] = bands.get('alpha', 0) / bands.get('beta', 0)
        else:
            ratios['alpha_beta'] = 0.0
        
        # Theta/Alpha ratio (indicates drowsiness)
        if bands.get('alpha', 0) > 0:
            ratios['theta_alpha'] = bands.get('theta', 0) / bands.get('alpha', 0)
        else:
            ratios['theta_alpha'] = 0.0
        
        # Beta/Delta ratio (indicates cognitive activity)
        if bands.get('delta', 0) > 0:
            ratios['beta_delta'] = bands.get('beta', 0) / bands.get('delta', 0)
        else:
            ratios['beta_delta'] = 0.0
        
        # (Alpha + Theta) / Beta ratio (fatigue indicator)
        if bands.get('beta', 0) > 0:
            ratios['alpha_theta_beta'] = (bands.get('alpha', 0) + bands.get('theta', 0)) / bands.get('beta', 0)
        else:
            ratios['alpha_theta_beta'] = 0.0
        
        return ratios
    
    def _get_empty_bands(self) -> Dict[str, float]:
        """Return empty frequency bands dictionary"""
        return {band: 0.0 for band in self.frequency_bands.keys()}
    
    def reset(self):
        """Reset the signal buffer"""
        self.raw_buffer = []
